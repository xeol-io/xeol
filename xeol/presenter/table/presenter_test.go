package table

import (
	"bytes"
	"flag"
	"strings"
	"testing"
	"time"

	"github.com/anchore/go-testutils"
	syftPkg "github.com/anchore/syft/syft/pkg"
	"github.com/anchore/syft/syft/source"
	"github.com/sergi/go-diff/diffmatchpatch"
	"github.com/stretchr/testify/assert"

	"github.com/xeol-io/xeol/xeol/eol"
	"github.com/xeol-io/xeol/xeol/match"
	"github.com/xeol-io/xeol/xeol/pkg"
	"github.com/xeol-io/xeol/xeol/presenter/models"
)

var update = flag.Bool("update", false, "update the *.golden files for table presenters")

func TestCreateRow(t *testing.T) {
	pkg := pkg.Package{
		ID:      "package-1-id",
		Name:    "package-1",
		Version: "1.0.1",
		Type:    syftPkg.DebPkg,
	}
	match1 := match.Match{
		Cycle: eol.Cycle{
			ProductName:       "MongoDB Server",
			ReleaseDate:       "2018-07-31",
			ReleaseCycle:      "2.8",
			Eol:               "2018-07-31",
			LatestReleaseDate: "2018-07-31",
		},
		Package: pkg,
	}
	match2 := match.Match{
		Cycle: eol.Cycle{
			ProductName:       "MongoDB Server",
			ReleaseDate:       "2018-07-31",
			ReleaseCycle:      "2.8",
			Eol:               "2025-01-01",
			LatestReleaseDate: "2018-07-31",
		},
		Package: pkg,
	}

	cases := []struct {
		name           string
		match          match.Match
		severitySuffix string
		expectedErr    error
		expectedRow    []string
	}{
		{
			name:        "create row for eol",
			match:       match1,
			expectedErr: nil,
			expectedRow: []string{match1.Package.Name, match1.Package.Version, match1.Cycle.Eol, "1614", match1.Package.Type.PackageURLType()},
		},
		{
			name:        "create row for eol in the future",
			match:       match2,
			expectedErr: nil,
			expectedRow: []string{match2.Package.Name, match2.Package.Version, match2.Cycle.Eol, "-", match2.Package.Type.PackageURLType()},
		},
	}

	now = func() time.Time { return time.Date(2022, 12, 31, 0, 0, 0, 0, time.UTC) }

	for _, testCase := range cases {
		t.Run(testCase.name, func(t *testing.T) {
			row, err := createRow(testCase.match)

			assert.Equal(t, testCase.expectedErr, err)
			assert.Equal(t, testCase.expectedRow, row)
		})
	}
}

func TestTablePresenter(t *testing.T) {

	var buffer bytes.Buffer
	matches, packages, _, _, _ := models.GenerateAnalysis(t, source.ImageScheme)

	pb := models.PresenterConfig{
		Matches:  matches,
		Packages: packages,
	}

	pres := NewPresenter(pb)

	// run presenter
	err := pres.Present(&buffer)
	if err != nil {
		t.Fatal(err)
	}
	actual := buffer.Bytes()
	if *update {
		testutils.UpdateGoldenFileContents(t, actual)
	}

	var expected = testutils.GetGoldenFileContents(t)

	if !bytes.Equal(expected, actual) {
		dmp := diffmatchpatch.New()
		diffs := dmp.DiffMain(strings.TrimSuffix(strings.TrimSpace(string(expected)), "\n"), strings.TrimSuffix(strings.TrimSpace(string(actual)), "\n"), true)
		t.Errorf("mismatched output:\n%s", dmp.DiffPrettyText(diffs))
	}

	// TODO: add me back in when there is a JSON schema
	// validateAgainstDbSchema(t, string(actual))
}

func TestEmptyTablePresenter(t *testing.T) {
	// Expected to have no output

	var buffer bytes.Buffer

	matches := match.NewMatches()

	pb := models.PresenterConfig{
		Matches:  matches,
		Packages: nil,
	}

	pres := NewPresenter(pb)

	// run presenter
	err := pres.Present(&buffer)
	if err != nil {
		t.Fatal(err)
	}
	actual := buffer.Bytes()
	if *update {
		testutils.UpdateGoldenFileContents(t, actual)
	}

	var expected = testutils.GetGoldenFileContents(t)

	if !bytes.Equal(expected, actual) {
		dmp := diffmatchpatch.New()
		diffs := dmp.DiffMain(string(expected), string(actual), true)
		t.Errorf("mismatched output:\n%s", dmp.DiffPrettyText(diffs))
	}

}
