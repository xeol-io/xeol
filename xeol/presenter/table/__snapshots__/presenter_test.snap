
[TestTablePresenter - 1]
[]uint8{0x4e, 0x41, 0x4d, 0x45, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x56, 0x45, 0x52, 0x53, 0x49, 0x4f, 0x4e, 0x20, 0x20, 0x45, 0x4f, 0x4c, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x44, 0x41, 0x59, 0x53, 0x20, 0x45, 0x4f, 0x4c, 0x20, 0x20, 0x54, 0x59, 0x50, 0x45, 0x20, 0xa, 0x70, 0x61, 0x63, 0x6b, 0x61, 0x67, 0x65, 0x2d, 0x31, 0x20, 0x20, 0x31, 0x2e, 0x31, 0x2e, 0x31, 0x20, 0x20, 0x20, 0x20, 0x32, 0x30, 0x31, 0x38, 0x2d, 0x30, 0x37, 0x2d, 0x33, 0x31, 0x20, 0x20, 0x31, 0x36, 0x31, 0x34, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x72, 0x70, 0x6d, 0x20, 0x20, 0x20, 0xa, 0x70, 0x61, 0x63, 0x6b, 0x61, 0x67, 0x65, 0x2d, 0x32, 0x20, 0x20, 0x32, 0x2e, 0x32, 0x2e, 0x32, 0x20, 0x20, 0x20, 0x20, 0x59, 0x45, 0x53, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x2d, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x64, 0x65, 0x62, 0x20, 0x20, 0x20, 0xa, 0x75, 0x62, 0x75, 0x6e, 0x74, 0x75, 0x20, 0x20, 0x20, 0x20, 0x20, 0x31, 0x36, 0x2e, 0x30, 0x34, 0x20, 0x20, 0x20, 0x20, 0x32, 0x30, 0x32, 0x31, 0x2d, 0x30, 0x34, 0x2d, 0x30, 0x32, 0x20, 0x20, 0x36, 0x33, 0x38, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x6f, 0x73, 0x20, 0x20, 0x20, 0x20, 0xa}
---

[TestEmptyTablePresenter - 1]
[]uint8{0xe2, 0x9c, 0x85, 0x20, 0x6e, 0x6f, 0x20, 0x45, 0x4f, 0x4c, 0x20, 0x73, 0x6f, 0x66, 0x74, 0x77, 0x61, 0x72, 0x65, 0x20, 0x68, 0x61, 0x73, 0x20, 0x62, 0x65, 0x65, 0x6e, 0x20, 0x66, 0x6f, 0x75, 0x6e, 0x64, 0xa}
---
